sdk/testing/stepwise/stepwise.go

```diff
@@ -0,0 +1,340 @@
+	}
+}
+
+// MountOptions are a collection of options each step driver should
+// support
+type MountOptions struct {
+	// MountPathPrefix is an optional prefix to use when mounting the plugin. If
+	// omitted the mount path will default to the PluginName with a random suffix.
+	MountPathPrefix string
+
+	// Name is used to register the plugin. This can be arbitrary but should be a
+	// reasonable value. For an example, if the plugin in test is a secret backend
+	// that generates UUIDs with the name "vault-plugin-secrets-uuid", then "uuid"
+	// or "test-uuid" would be reasonable. The name is used for lookups in the
+	// catalog. See "name" in the "Register Plugin" endpoint docs:
+	// - https://www.vaultproject.io/api-docs/system/plugins-catalog#register-plugin
+	RegistryName string
+
+	// PluginType is the optional type of plugin. See PluginType const defined
+	// above
+	PluginType PluginType
+
+	// PluginName represents the name of the plugin that gets compiled. In the
+	// standard plugin project file layout, it represents the folder under the
+	// cmd/ folder. In the below example UUID project, the PluginName would be
+	// "uuid":
+	//
+	// vault-plugin-secrets-uuid/
+	// - backend.go
+	// - cmd/
+	// ----uuid/
+	// ------main.go
+	// - path_generate.go
+	//
+	PluginName string
+}
+
+// Step represents a single step of a test Case
+type Step struct {
+	// Operation defines what action is being taken in this step; write, read,
+	// delete, et. al.
+	Operation Operation
+
+	// Path is the localized request path. The mount prefix, namespace, and
+	// optionally "auth" will be automatically added.
+	Path string
+
+	// Arguments to pass in the request. These arguments represent payloads sent
+	// to the API.
+	Data map[string]interface{}
+
+	// Assert is a function that is called after this step is executed in order to
+	// test that the step executed successfully. If this is not set, then the next
+	// step will be called
+	Assert AssertionFunc
+
+	// Unauthenticated will make the request unauthenticated.
+	Unauthenticated bool
+}
+
+// AssertionFunc is the callback used for Assert in Steps.
+type AssertionFunc func(*api.Secret, error) error
+
+// Case represents a scenario we want to test which involves a series of
+// steps to be followed sequentially, evaluating the results after each step.
+type Case struct {
+	// Environment is used to setup the Vault instance and provide the client that
+	// will be used to drive the tests
+	Environment Environment
+
+	// Steps are the set of operations that are run for this test case. During
+	// execution each step will be logged to output with a 1-based index as it is
+	// ran, with the first step logged as step '1' and not step '0'.
+	Steps []Step
+
+	// SkipTeardown allows the Environment TeardownFunc to be skipped, leaving any
+	// infrastructure created after the test exists. Depending on the Environment
+	// used this could incur costs the user is responsible for.
+	SkipTeardown bool
```

Comment:
Are there situations where `SkipTeardown:true` and `Environment.Teardown ` is not a no-op. If not, then a client can just have `Teardown` perform a no-op instead of requiring an additional bool field.

Reply:
Environments are responsible for knowing how to destroy and/or clean up any infrastructure they created during testing. The purpose of `SkipTeardown` is for developers to preserve the environment for inspection after a test runs, for whatever debugging reasons they may have. I've update the code comment above in 4951eabb5 to mention this, does that clarify things here? Or perhaps I'm not following your suggestion/question correctly?

Reply:
Cool, makes complete sense. Thanks!